### PR TITLE
feat(state): prune block on commit

### LINKED ISSUE #1393: Pruning on New Block 1:1
Once a new block is committed, the following operations should be performed:

1. If the node is a Full Node, no action is needed.
2. If the node is a Pruned Node, call the PruneBlock function with the block number equal to LastBlockHeight - RetentionBlocks.

This process ensures that for each new block added, one old block will be removed, keeping the store blocks up to the RetentionBlocks number.

https://pips.pactus.org/PIPs/pip-27#pruning-on-new-block

### PR DESCRIPTION
**Pull request by kehiy**
## Description

This PR adds prune on a new block for pruned nodes.

## Related issue
- Fixes #1393

### CODE REVIEW

File: store/interface.go [hunk 1]
```diff
@@ -99,6 +99,7 @@ type Reader interface {
 	LastCertificate() *certificate.BlockCertificate
 	IsBanned(addr crypto.Address) bool
 	Prune(resultFunc func(pruned, skipped, pruningHeight uint32)) error
+	PruneOnCommit() error
```
Comment (b00f) [inline store/interface.go:102]: We don't need to tell the store to prune the block; the store should know when to prune or not. Let each module do its job

File: store/store.go [hunk 2]
```diff
@@ -425,7 +434,30 @@ func (s *store) Prune(resultFunc func(pruned, skipped, pruningHeight uint32)) er
 	return nil
 }
 
-func (s *store) pruneBlock(blockHeight uint32) (bool, error) { //nolint
+func (s *store) PruneOnCommit() error {
```
Comment (b00f) [inline store/store.go:437]: This logic should move to commit block function.

**Review by b00f — CHANGES_REQUESTED** [review]

File: store/store.go (hunk 2 above)
Comment (kehiy) [reply store/store.go:437]: But we have `RetentionBlocks` in the store, and `isPruned` can be determined properly in the store only.

Any way to have them in state as well without issues?

Comment (b00f) [reply store/store.go:437]: We can export a function, `IsPruned`, to inform the outside world that the store is in Prune mode. Remember, based on PIP-27, `pruneBlock` is private, so pruning occurs inside the store.

File: store/store.go [hunk 3]
```diff
@@ -147,6 +155,24 @@ func (s *store) SaveBlock(blk *block.Block, cert *certificate.BlockCertificate)
...
+		if lc == nil {
+			return
+		}
+
+		pruneH := lc.Height() - s.config.RetentionBlocks()
+		if pruneH <= 0 {
+			return
+		}
+
+		deleted, err := s.pruneBlock(pruneH)
```
Comment (Ja7ad) [inline store/store.go:170]: After calling `pruneBlock` you need the call method `WriteBatch`

File: store/store.go [hunk 4]
```diff
@@ -93,13 +94,20 @@ func NewStore(conf *Config) (Store, error) {
 		txStore:        newTxStore(db, conf.TxCacheSize),
 		accountStore:   newAccountStore(db, conf.AccountCacheSize),
 		validatorStore: newValidatorStore(db),
+		isPruned:       false,
```
Comment (Ja7ad) [inline store/store.go:97]: Don't require to set the default value, because the default value boolean is false

**Review by Ja7ad — CHANGES_REQUESTED** [review]

File: store/store.go (hunk 3 above)
Comment (b00f) [reply store/store.go:170]: @Ja7ad  It is called externally.

File: store/store.go (hunk 4 above)
Comment (b00f) [reply store/store.go:97]: It's okay. Let it be. Some linters are sensitive about not setting values explicitly.

File: store/interface.go [hunk 5]
```diff
@@ -107,6 +106,7 @@ type Store interface {
 	UpdateAccount(addr crypto.Address, acc *account.Account)
 	UpdateValidator(val *validator.Validator)
 	SaveBlock(blk *block.Block, cert *certificate.BlockCertificate)
+	Prune(resultFunc func(pruned, skipped, pruningHeight uint32)) error
```
Comment (b00f) [inline store/interface.go:109]: Remove this function, and add `IsPrunned() bool`

Comment (b00f) [inline store/interface.go:109]: Remove this function

Comment (kehiy) [reply store/interface.go:109]: I just moved it no added it.

File: store/interface.go [hunk 6]
```diff
@@ -98,7 +98,6 @@ type Reader interface {
 	TotalValidators() int32
 	LastCertificate() *certificate.BlockCertificate
 	IsBanned(addr crypto.Address) bool
-	Prune(resultFunc func(pruned, skipped, pruningHeight uint32)) error
```
[the remark this answers is not in the document]
Comment (kehiy) [reply store/interface.go:101]: Done

**Review by b00f — APPROVED** [review]

**Review by Ja7ad — APPROVED** [review]

**Pull request merged**